### ISSUE TITLE
Q: Dynamic node size depending on zoom level

### ISSUE THREAD

**Issue by kaij** [jacomyal/sigma.js#1198]
I'm visualizing a network of persons with Sigma - which works very nicely. These networks can be anything in size from very small (1-5) to very large (thousands of nodes).

Currently, the node size seems to stay fixed, independent on zoom level. When using a fixed nodesize on graph creation, the nodes either the nodes overlap (when zoomed out), or the nodes are too small (when zoomed in). See examples: 
![image](https://user-images.githubusercontent.com/1143418/148643024-d55a8454-6aaf-4378-8ffa-1cd20d19374f.png) 
![image](https://user-images.githubusercontent.com/1143418/148643040-f48e70a4-a493-4788-8c52-7578998c3402.png)

Also, when zoomed in, the image does not grow above its original size (although the label does). I would like the node to get bigger when zoomed in. Maybe it wouldn't grow linearly but slightly less when zoomed in.  

Is there a way to tweak this?

**Comment by kaij** [comment]
Illustration: Label grows, but image stays the same size
![image](https://user-images.githubusercontent.com/1143418/148643424-ec0c44ae-8313-4f87-96e3-61c488b35dcd.png)

**Comment by jacomyal** [comment]
Hi @kaij,

There is a specific design choice in sigma, which makes node radius and edge thickness proportional to **the zoom ratio's square root** instead of the zoom ratio itself. That makes nodes not too small when zoomed out, and not too big when zoomed in.

But in your case, it looks like something is going wrong.
To help us fix this issue, can you please:
- Try [the existing examples](https://github.com/jacomyal/sigma.js/#examples) to see if this issue occurs as well?
- If it does not occur on existing examples, try to produce a minimal reproduction case?

Thanks in advance !

**Comment by kaij** [comment]
Hi @jacomyal, thanks for the quick response. 

Yes, I saw that you mentioned the square root ratio in #1185. But that would mean the node size should still grow when zooming in (as the edge size does) if I understand correctly. 

All the examples (as well as my visualization) however seem to limit the node size to 32px or 64px, i.e. a node is never bigger than 32px in my case, no matter how deep I zoom in. The screenshot below is from this example: https://codesandbox.io/s/github/jacomyal/sigma.js/tree/main/examples/custom-rendering.
![image](https://user-images.githubusercontent.com/1143418/149166151-e7ca158a-1f74-4f1b-add1-39873580447e.png)

I have no idea where the 32px limit comes from (texture size?) and also tried to fiddle around with the node.image program, but it didn't have any effect. Any input on what I could check/enhance/fix is warmly welcome, at the moment I'm still struggling with understanding the different modules in SigmaJS and how they play together.

**Comment by jacomyal** [comment]
But when you zoom out for long enough, do the nodes sizes decrease at some point?

Also, can you please tell me:
1. What browser do you use?
2. What device do you use?
3. What does `window.devicePixelRatio` returns, in your browser's JavaScript console?

I'm also thinking about something else, by the way:
The [`node.fast`](https://github.com/jacomyal/sigma.js/blob/e187ba5b567eb359f2b4e438987785da73684667/src/rendering/webgl/programs/node.fast.ts#L58) and [`node.image`](https://github.com/jacomyal/sigma.js/blob/923b26e2c2ffb4ec85f57747d2aa8d1fd0ee1390/src/rendering/webgl/programs/node.image.ts#L249) programs both use the `gl.POINTS` drawing method, which means we "carve" the node disc in a single square determined by a single point position, when the [`node`](https://github.com/jacomyal/sigma.js/blob/e187ba5b567eb359f2b4e438987785da73684667/src/rendering/webgl/programs/node.ts#L142) program uses the `gl.TRIANGLES` method, in which we carve the disc in a triangle determined by three different positions.

The first method scales way more, because we have much less data to send to the graphic card (more than three times less), but in the meantime the `gl.POINTS` method has a maximum size that it won't exceed (but that is more about 275px or something, for me):

![image](https://user-images.githubusercontent.com/144846/149172326-fae9ca10-8d8b-451d-baec-c56567b32cda.png)

Can you try to use the `node` program, to see if the problem persists?

**Comment by kaij** [comment]
Yes, on zooming out, node size decreases.

I tested (node.image) with the following configurations:
1. Macbook Pro M1 / macOS: window.devicePixelRatio = 2 (internal display), window.devicePixelRatio = 1 (external display). Browsers Chrome, Firefox, Safari. All Browsers: limited node size to 32px (internal display) or 64px (external display) 
2. Windows 11 Notebook: window.devicePixelRatio = 1.25. Browsers Edge/Chromium, Firefox. Both browsers limited node size to 1024px (!)

I also tested with the node program, but there I get the effect that the node size is decreasing continuously when zooming in (I guess towards a size of 1px) - see image.

Just to make sure, I used the following code for the node program (simplified):
```
import NodeProgram from 'sigma/rendering/webgl/programs/node';
const renderer = new Sigma(graph, container, { defaultNodeType: 'node', nodeProgramClasses: { node: NodeProgram } });
```

So it seems like a macOS/WebGL2 platform issue. Any idea on how to approach this best?

![image](https://user-images.githubusercontent.com/1143418/149211798-a0f9376a-4b5b-473e-8c61-4c7c0e91a424.png)

**Comment by kaij** [comment]
Just a quick update on this: using the node program (with the triangles) works fine for the scaling (I moved the program to my own code, now it works - it seems to have been a loader problem). What I'm trying now is to combine node(triangles), border and image program into one single program so it will look nicer.

**Comment by kaij** [comment]
@jacomyal I have now merged the node and node.image program into a new program which positions and draws the circle textures using the triangle method. Image loading is more or less a copy of the node.image program. It's currently locally but I'd happy to provide a pull request if you agree. If you want, I can also merge it directly into the node program. What do you think?

**Comment by jacomyal** [comment]
Nice, thanks! I would like very much indeed, since I'm sure people (including myself) can need this sometime. Also, since this program scales less than the other (mostly because using the triangle method requires having a way longer byte array), I'd prefer if it does not replace any existing program, but comes as a new one. In this case, do you have ideas on how to name it?

Also, note to myself, for #1146, I worked a bit on the `sigma.min.js` bundle creation, and it seems to embed all programs now, without actually exposing them. I think we'll need to properly build each program as an independant minified JS file.

**Comment by kaij** [comment]
@jacomyal What about `node.combined`? - mostly because in the end it would be nice to also integrate drawing a border in the shader (image, border, triangle). I'm open - you can decide about the name on what's best for sigma.

**Comment by jacomyal** [comment]
This ticket is still relevant, sorry I hadn't enough time on sigma recently...

Also I somehow missed your last message @kaij, sorry about this. I think `node.combined` is a good name indeed.

**Comment by simonavril** [comment]
I have also experienced this issue and just couldn't figure it out for a while. Eventually found at that using gl.POINTS has a hard limit on how big the square can be drawn (square that is transformed into a circle in the fragment shader).

To find out the range of acceptable size for the node you can use:
const [minSize, maxSize] = gl.getParameter(gl.ALIASED_POINT_SIZE_RANGE);

and for me it maxes at 255 which means after a while the zoom ends up not working anymore and the node ends up being smaller and smaller than the canvas hovered node

**Comment by TGOlson** [comment]
> #1207 fix canvas size calculation #1209

I am quite late here, but also finding image scaling to be a problem with the default node image problem. 

I tested the fix from @kaij in https://github.com/jacomyal/sigma.js/pull/1206 and it does indeed fix the image scaling problem.

However, I think there is a problem with the solution from @kaij in that node sizing doesn't seem to work correctly (or at least, something like hit detection for the nodes?). 

Basically, the labels and hover detection seem to treat the node as very small, even when it's scaled up to a large size. The result is that when zoomed in the labels are hidden behind the node, and hover detection only works if the mouse is over the very center of the node. 

Again, I realize this is quite an outdated thread, but wondering if anyone had pointers towards a potential fix. Thanks!

**Comment by RouxRC** [comment]
@TGOlson I believe this point is fixed within the current [WIP v3](https://github.com/jacomyal/sigma.js/tree/v3.0.0-alpha3) which you can already try to play with as there is a prerelease of it within npm.

**Comment by TGOlson** [comment]
Thanks @RouxRC, I think I found the issue. It was with how this fix was computing `size`, which seemed to render the nodes larger, but in turn broke the hover/labels.

I briefly looked at the WIP v3 branch, but the main thing I need is `gl.TRIANGLES` rendering for images so that they scale well when enlarged.
